Handling commands to unknown orders

ale64bit opened the pull request:
Created a new event for `Order`, in order to avoid manipulating uninitialized orders. After some configurable threshold, the uninitialized orders automatically stop themselves in order to not litter the cluster with unnecessary actors. Also the initialization command is sent from the command handler when the order is created, or even if the order was already registered in the `Orders` singleton, to avoid consistency issues when restarting the system, entry relocation, network partition, etc. Added some tests to check correctness also.

jgordijn commented on src/main/scala/cqrs/settings/Settings.scala:
```diff
@@ -12,6 +12,7 @@ class Settings(system: ExtendedActorSystem) extends Extension {
   val config = system.settings.config getConfig ("cqrs")
 
   val maxOrderPrice: Double = config getDouble "maxOrderPrice"
+  val uninitializedOrderLifeTime: Duration = config getInt "uninitializedOrderLifeTime" seconds
```
Typesafe config supports duration specification in config. This allows more flexibility in specifying the time.

``` scala
import scala.concurrent.duration.{ Duration, MILLISECONDS }
val uninitializedOrderLifeTime: Duration = Duration(system.settings.config.getDuration("uninitializedOrderLifeTime", MILLISECONDS), MILLISECONDS)
```

and then in the config (see: https://github.com/typesafehub/config/blob/master/HOCON.md#duration-format): 

```
uninitializedOrderLifeTime = 30 seconds 
```

jgordijn commented on src/main/scala/cqrs/settings/Settings.scala:
```diff
@@ -1,7 +1,7 @@
 
 import akka.actor.{ Actor, ExtensionKey, Extension, ExtendedActorSystem }
-import scala.concurrent.duration.{ Duration, FiniteDuration }
+import scala.concurrent.duration._
```
I try to follow: http://twitter.github.io/effectivescala/#Formatting-Imports

_Use wildcards when more than six names are imported_

jgordijn commented on src/main/scala/cqrs/domain/orders/Order.scala:
```diff
@@ -42,19 +49,17 @@ class Order(maxOrderPrice: Double) extends PersistentActor with SettingsActor wi
+        log.debug(s"Order initialized {}", persistenceId)
+        updateState(evt)
+      }
+    case ReceiveTimeout ⇒ context stop self
```
Passivate (http://doc.akka.io/docs/akka/2.3.9/contrib/cluster-sharding.html#Passivation) would be the right option here.

We can also safely passivate everywhere when a `ReceiveTimeout` is received.

jgordijn commented on src/main/scala/cqrs/domain/orders/Orders.scala:
```diff
@@ -42,10 +42,12 @@ class Orders extends PersistentActor with SettingsActor with ActorLogging {
       log.debug("Creating Order")
       persist(OrderCreated(orderId, username)) { evt ⇒
         updateState(evt)
+        orderRegion ! OrderCommandHandler.Command(orderId, Order.InitializeOrder)
```
Consider the following scenario:
Message sent to Orders (persisted)
Message sent to Order does not arrive (due to network problem, or persistence problem or something else).

`Orders` knows the `Order`, but the `Order` is not initialized. It is not possible to initialize the order anymore. We need 2 things here: 
1. `Order` needs to be idempotent. Allowing to be initialized multiple times and Acking every time it is initialized. Initializing after it already is initialized, does not really change anything in the `Order`. No event needs to be persisted.
2. `Orders` only persist `OrderCreated` when the `Order` Acknowledge that it is initialized

If we do it this way, we make it fool proof. `Orders` will only know initialized `Order`s.

jgordijn commented on src/test/scala/cqrs/domain/orders/DomainTest.scala:
```diff
@@ -2,6 +2,8 @@ package cqrs
 package domain.orders
 
 import java.util.UUID
+import cqrs.domain.orders.OrderCommandHandler.UnknownOrderException
```
this import needs to be on line 18 and you can change it to:
`import OrderCommandHandler.UnknownOrderException`

jgordijn commented on src/main/scala/cqrs/domain/orders/Order.scala:
```diff
@@ -1,13 +1,15 @@
-import akka.actor.{ Status, ActorLogging, Props }
+import akka.actor.{ReceiveTimeout, Status, ActorLogging, Props}
 import akka.persistence.PersistentActor
+import cqrs.domain.orders.OrderCommandHandler.UnknownOrderException
```
You can change `import cqrs.domain.orders.OrderCommandHandler.UnknownOrderException`
to `import OrderCommandHandler.UnknownOrderException`

jgordijn commented:
Thank you very much for contributing! I really appreciate it! I've added a few remarks.

ale64bit commented:
You are welcome) Btw, it's a very good template.
Changed a few things, regarding the remarks. Good point about the consistency issue for the Ack between `Order` and `Orders`. Added the Ack message and `Orders` only persist the `OrderCreated` event after receiving Ack from the initialized order (is it better to just `ask`, instead of creating a new message?). Also added passivation for the `ReceiveTimeout`s.

jgordijn commented:
Using `ask` will introduce extra complexity. You will have to send a message to yourself as `persist` is not thread safe so you cannot call it in a callback hook on `ask`.

Great stuff! Thanks again!
